File: prow/spyglass/lenses/coverage/coverage.ts
/*
Copyright 2019 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Color from "color";
import {Coverage, parseCoverage} from 'io_k8s_test_infra/gopherage/cmd/html/static/parser';
import {inflate} from "pako/lib/inflate";

declare const COVERAGE_FILE: string;
declare const RENDERED_COVERAGE_URL: string;

const NO_COVERAGE = Color('#FF0000');
const FULL_COVERAGE = Color('#00FF00');

// Inspired by https://dl.acm.org/citation.cfm?id=949654
function renderChildren(parent: Node, coverage: Coverage, horizontal: boolean): void {
  let offset = 0;
  for (const child of coverage.children.values()) {
    const node = document.createElement('a');
    node.style.display = 'block';
    parent.appendChild(node);
    const percentage = child.totalStatements / coverage.totalStatements * 100;
    node.style.position = 'absolute';
    if (horizontal) {
      node.style.width = `${percentage}%`;
      node.style.height = '100%';
      node.style.top = '0';
      node.style.left = `${offset}%`;
    } else {
      node.style.width = '100%';
      node.style.height = `${percentage}%`;
      node.style.top = `${offset}%`;
      node.style.left = `0`;
    }
    offset += percentage;
    if (child.totalFiles === 1) {
      node.classList.add('leaf');
      const [filename, file] = child.files.entries().next().value;
      node.title = `${filename}: ${(file.coveredStatements / file.totalStatements * 100).toFixed(0)}%`;

      const bgColor = NO_COVERAGE.mix(FULL_COVERAGE, file.coveredStatements / file.totalStatements);
      node.style.backgroundColor = bgColor.hex();
      // Not having a border looks weird, but using a constant colour causes tiny boxes
      // to consist entirely of that colour. By using a border colour based on the
      // box colour, we still show some information.
      node.style.borderColor = bgColor.darken(0.3).hex();

      if (RENDERED_COVERAGE_URL) {
        node.href = `${RENDERED_COVERAGE_URL}#file${file.fileNumber}`;
      }
    } else {
      renderChildren(node, child, !horizontal);
    }
  }
}

window.onload = () => {
  // Because the coverage files are a) huge, and b) compress excellently, we send it as
  // gzipped base64. This is faster unless your internet connection is faster than
  // about 300 Mb/s.
  const content = inflate(atob(COVERAGE_FILE), {to: 'string'});
  const coverage = parseCoverage(content);
  document.getElementById('statement-coverage')!.innerText = `${(coverage.coveredStatements / coverage.totalStatements * 100).toFixed(0)}% (${coverage.coveredStatements.toLocaleString()} of ${coverage.totalStatements.toLocaleString()} statements)`;
  document.getElementById('file-coverage')!.innerText = `${(coverage.coveredFiles / coverage.totalFiles * 100).toFixed(0)}% (${coverage.coveredFiles.toLocaleString()} of ${coverage.totalFiles.toLocaleString()} files)`;
  const treemapEl = document.getElementById('treemap')!;
  renderChildren(treemapEl, coverage, true);
  if (RENDERED_COVERAGE_URL) {
    treemapEl.classList.add('interactive');
  }
};

File: prow/spyglass/lenses/coverage/coverage.go
/*
Copyright 2019 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package coverage provides a coverage viewer for Spyglass
package coverage

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"html/template"
	"path/filepath"
	"strings"

	"github.com/sirupsen/logrus"

	"k8s.io/test-infra/prow/config"
	"k8s.io/test-infra/prow/spyglass/api"
	"k8s.io/test-infra/prow/spyglass/lenses"
)

const (
	name     = "coverage"
	title    = "Coverage"
	priority = 7
)

func init() {
	lenses.RegisterLens(Lens{})
}

// Lens is the implementation of a coverage-rendering Spyglass lens.
type Lens struct{}

// Config returns the lens's configuration.
func (lens Lens) Config() lenses.LensConfig {
	return lenses.LensConfig{
		Name:     name,
		Title:    title,
		Priority: priority,
	}
}

// Header renders the content of <head> from template.html.
func (lens Lens) Header(artifacts []api.Artifact, resourceDir string, config json.RawMessage, spyglassConfig config.Spyglass) string {
	t, err := template.ParseFiles(filepath.Join(resourceDir, "template.html"))
	if err != nil {
		return fmt.Sprintf("<!-- FAILED LOADING HEADER: %v -->", err)
	}
	var buf bytes.Buffer
	if err := t.ExecuteTemplate(&buf, "header", nil); err != nil {
		return fmt.Sprintf("<!-- FAILED EXECUTING HEADER TEMPLATE: %v -->", err)
	}
	return buf.String()
}

// Callback does nothing.
func (lens Lens) Callback(artifacts []api.Artifact, resourceDir string, data string, config json.RawMessage, spyglassConfig config.Spyglass) string {
	return ""
}

// Body renders the <body>
func (lens Lens) Body(artifacts []api.Artifact, resourceDir string, data string, config json.RawMessage, spyglassConfig config.Spyglass) string {
	if len(artifacts) == 0 {
		logrus.Error("coverage Body() called with no artifacts, which should never happen.")
		return "Why am I here? There is no coverage file."
	}

	profileArtifact := artifacts[0]
	var htmlArtifact api.Artifact
	if len(artifacts) > 1 {
		if len(artifacts) > 2 {
			return "Too many files - expected one coverage file and one optional HTML file"
		}
		if strings.HasSuffix(artifacts[0].JobPath(), ".html") {
			htmlArtifact = artifacts[0]
			profileArtifact = artifacts[1]
		} else if strings.HasSuffix(artifacts[1].JobPath(), ".html") {
			htmlArtifact = artifacts[1]
			profileArtifact = artifacts[0]
		} else {
			return "Multiple input files, but none had a .html extension."
		}
	}

	content, err := profileArtifact.ReadAll()
	if err != nil {
		logrus.WithError(err).Warn("Couldn't read a coverage file that should exist.")
		return fmt.Sprintf("Faiiled to read the coverage file: %v", err)
	}

	coverageTemplate, err := template.ParseFiles(filepath.Join(resourceDir, "template.html"))
	if err != nil {
		logrus.WithError(err).Error("Error executing template.")
		return fmt.Sprintf("Failed to load template file: %v", err)
	}

	w := &bytes.Buffer{}
	g := gzip.NewWriter(w)
	_, err = g.Write(content)
	if err != nil {
		logrus.WithError(err).Warn("Failed to compress coverage file")
		return fmt.Sprintf("Failed to compress coverage file: %v", err)
	}
	if err := g.Close(); err != nil {
		logrus.WithError(err).Warn("Failed to close gzip for coverage file")
		return fmt.Sprintf("Failed to close gzip for coverage file: %v", err)
	}
	result := base64.StdEncoding.EncodeToString(w.Bytes())

	renderedCoverageURL := ""
	if htmlArtifact != nil {
		renderedCoverageURL = htmlArtifact.CanonicalLink()
	}
	t := struct {
		CoverageContent  string
		RenderedCoverage string
	}{
		CoverageContent:  result,
		RenderedCoverage: renderedCoverageURL,
	}
	var buf bytes.Buffer
	if err := coverageTemplate.ExecuteTemplate(&buf, "body", t); err != nil {
		logrus.WithError(err).Error("Error executing template.")
	}

	return buf.String()
}
